sa/metrics.go

```diff
@@ -0,0 +1,125 @@
+// stats will be updated based on the gorp dbMap's inner sql.DBMap's DBStats
+// structure values.
+func InitDBMetrics(dbMap *gorp.DbMap, scope metrics.Scope) {
+	// Create a dbMetrics instance and register prometheus metrics
+	dbm := newDbMetrics(dbMap, scope)
+
+	// Start the metric reporting goroutine to update the metrics periodically.
+	go dbm.reportDBMetrics()
+}
+
+// newDbMetrics constructs a dbMetrics instance by registering prometheus stats.
+func newDbMetrics(dbMap *gorp.DbMap, scope metrics.Scope) *dbMetrics {
+	maxOpenConns := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "max_open_connections",
+		Help: "Maximum number of DB connections allowed.",
+	})
+	scope.MustRegister(maxOpenConns)
+
+	openConns := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "open_connections",
+		Help: "Number of established DB connections (in-use and idle).",
+	})
+	scope.MustRegister(openConns)
+
+	inUse := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "inuse",
+		Help: "Number of DB connections currently in use.",
+	})
+	scope.MustRegister(inUse)
+
+	idle := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "idle",
+		Help: "Number of idle DB connections.",
+	})
+	scope.MustRegister(idle)
+
+	waitCount := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "wait_count",
+		Help: "Total number of DB connections waited for.",
+	})
+	scope.MustRegister(waitCount)
+
+	waitDuration := prometheus.NewHistogram(prometheus.HistogramOpts{
+		Name: "wait_duration",
+		Help: "The total time blocked waiting for a new connection.",
+	})
+	scope.MustRegister(waitDuration)
+
+	maxIdleClosed := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "max_idle_closed",
+		Help: "Total number of connections closed due to SetMaxIdleConns.",
+	})
+	scope.MustRegister(maxIdleClosed)
+
+	maxLifetimeClosed := prometheus.NewCounter(prometheus.CounterOpts{
+		Name: "max_lifetime_closed",
+		Help: "Total number of connections closed due to SetConnMaxLifetime.",
+	})
+	scope.MustRegister(maxLifetimeClosed)
+
+	// Construct a dbMetrics instance with all of the registered metrics and the
+	// gorp DBMap
+	return &dbMetrics{
+		dbMap:              dbMap,
+		maxOpenConnections: maxOpenConns,
+		openConnections:    openConns,
+		inUse:              inUse,
+		idle:               idle,
+		waitCount:          waitCount,
+		waitDuration:       waitDuration,
+		maxIdleClosed:      maxIdleClosed,
+		maxLifetimeClosed:  maxLifetimeClosed,
+	}
+}
+
+// updateFrom updates the dbMetrics prometheus stats based on the provided
+// sql.DBStats object.
+func (dbm *dbMetrics) updateFrom(dbStats sql.DBStats) {
+	dbm.maxOpenConnections.Set(float64(dbStats.MaxOpenConnections))
+	dbm.openConnections.Set(float64(dbStats.OpenConnections))
+	dbm.inUse.Set(float64(dbStats.InUse))
+	dbm.idle.Set(float64(dbStats.InUse))
+	dbm.waitCount.Set(float64(dbStats.WaitCount))
+	dbm.waitDuration.Observe(dbStats.WaitDuration.Seconds())
```

Comment:
Looking at the docs I think this _might_ actually make more sense as a counter rather than a hist, since it's a single value we are adding each second. Thinking about how we'd observe this  I feel like we'd look at a rate of change... but not entirely sure.

Reply:
Good call, I agree.

Reply:
That makes sense to me. Implemented in 2368927. Thanks!